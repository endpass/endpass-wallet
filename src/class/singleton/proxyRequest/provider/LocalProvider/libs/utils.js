import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
