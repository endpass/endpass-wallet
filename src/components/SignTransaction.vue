<template>
  <div>
    <v-form
      id="signing-transaction-form"
      @submit="togglePasswordModal"
    >
      <v-textarea
        v-model="transaction"
        label="Transaction object (JSON)"
        placeholder="This is a transaction that you are signing to prove that you own it."
      />
      <v-button
        :disabled="!transaction"
        class-name="is-primary is-medium"
      >Sign transaction</v-button>
      <div
        v-if="signedTransaction"
        class="field"
      >
        <label class="label">RLP encoded transaction</label>
        <p class="code">{{ signedTransaction }}</p>
      </div>
    </v-form>
    <code-password-modal
      v-if="isPasswordModal"
      :code="transaction"
      label="Transaction object"
      @confirm="signTransaction"
      @close="togglePasswordModal"
    />
  </div>
</template>

<script>
import { mapGetters } from 'vuex';
import CodePasswordModal from '@/components/modal/CodePasswordModal';
import modalMixin from '@/mixins/modal';

export default {
  name: 'SignTransaction',

  data: () => ({
    transaction: '',
    signedTransaction: null,
  }),

  computed: {
    ...mapGetters('accounts', ['wallet']),
  },

  methods: {
    async signTransaction(password) {
      try {
        this.togglePasswordModal();

        this.signedTransaction = await this.wallet.signTransaction(
          JSON.parse(this.transaction),
          password,
        );
      } catch (error) {
        this.signedTransaction = null;
        this.$notify({
          title: 'Error signing transaction',
          text:
            'An error occurred while signing the transaction. Please try again.',
          type: 'is-danger',
        });
        console.error(error);
      }
    },
  },

  mixins: [modalMixin],

  components: {
    CodePasswordModal,
  },
};
</script>
