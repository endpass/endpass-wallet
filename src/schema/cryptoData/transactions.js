export const cryptoDataTransaction = {
  type: 'object',
  required: ['timestamp', 'from', 'to', 'hash', 'value'],
  properties: {
    data: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]+',
    },
    from: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]{40,}',
    },
    hash: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]{64,}',
    },
    timestamp: {
      type: 'number',
    },
    to: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]{40,}',
    },
    // token: {
    //   type: 'object',
    //   required: ['name', 'symbol'],
    //   properties: {
    //     name: {
    //       type: 'string',
    //     },
    //     symbol: {
    //       type: 'string',
    //     },
    //     decimals: {
    //       type: 'number',
    //       default: 18,
    //     },
    //   },
    // },
    value: {
      type: 'number',
    },
  },
};

export const cryptoDataTransactions = {
  type: 'array',
  items: cryptoDataTransaction,
};
