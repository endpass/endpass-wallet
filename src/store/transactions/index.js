import actions from './actions';
import getters from './getters';
import mutations from './mutations';

const state = {
  pendingTransactions: [],
  transactionHistory: [],
  pendingTransactionsFilterId: null,
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
