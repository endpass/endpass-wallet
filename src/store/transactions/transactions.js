import Tx from 'ethereumjs-tx';
import { BigNumber } from 'bignumber.js';
import { EventEmitter, NotificationError, Transaction } from '@/class';
import { MAIN_NET_ID } from '@/constants';
import ethplorerService from '@/services/ethplorer';
import web3 from '@/utils/web3';

export default {
  namespaced: true,
  state: {
    pendingTransactions: [],
    transactionHistory: [],
  },
  getters: {
    accountTransactions(state, getters, rootState) {
      if (!rootState.accounts.address) {
        return [];
      }

      let transactions = state.pendingTransactions;
      const { id: currentNetID } = rootState.web3.activeNet;

      if (currentNetID === MAIN_NET_ID) {
        transactions = transactions.concat(getters.filteredHistoryTransactions);
      }

      const address = rootState.accounts.address.getChecksumAddressString();

      return transactions
        .filter(trx => {
          const { to, from, state: trxStatus } = trx;

          return (
            web3.utils.toChecksumAddress(from) === address ||
            (trxStatus === 'success' &&
              web3.utils.toChecksumAddress(to) === address)
          );
        })
        .sort((trx1, trx2) => {
          if (typeof trx2.date === 'undefined') return 1;
          if (typeof trx1.date === 'undefined') return -1;
          return trx2.date - trx1.date;
        });
    },
    // Trx for the current network
    currentNetTransactions(state, getters, rootState) {
      const { id: currentNetID } = rootState.web3.activeNet;

      return getters.accountTransactions.filter(
        ({ networkId }) => networkId === currentNetID,
      );
    },
    // Exclude trx that exist in pendingTransactions
    filteredHistoryTransactions(state) {
      const isInPending = itemHash =>
        state.pendingTransactions.some(({ hash }) => hash === itemHash);

      return state.transactionHistory.filter(({ hash }) => !isInPending(hash));
    },
    pendingBalance(state, getters, rootState) {
      if (!rootState.accounts.address) {
        return '0';
      }

      const address = rootState.accounts.address.getChecksumAddressString();
      const networkId = rootState.web3.activeNet.id;

      return state.pendingTransactions
        .filter(
          tnx =>
            tnx.state === 'pending' &&
            tnx.from === address &&
            tnx.networkId === networkId,
        )
        .map(tnx => {
          const tnxValue = tnx.token === 'ETH' ? tnx.valueWei : '0';

          return BigNumber(tnx.gasCost).plus(tnxValue);
        })
        .reduce((total, item) => total.plus(item), BigNumber('0'))
        .toFixed();
    },
  },
  mutations: {
    addTransaction(state, transaction) {
      state.pendingTransactions.push(transaction);
    },
    updateTransaction(state, { hash, payload }) {
      const trxIndex = state.pendingTransactions.findIndex(
        trx => trx.hash === hash,
      );
      Object.assign(state.pendingTransactions[trxIndex], payload);
    },
    setTransactionHistory(state, transactions) {
      state.transactionHistory = transactions || [];
    },
  },
  actions: {
    async getNonceInBlock({ rootState }) {
      const address = rootState.accounts.address.getChecksumAddressString();

      return web3.eth.getTransactionCount(address);
    },
    async getNextNonce({ state, dispatch }) {
      const nonce = await dispatch('getNonceInBlock');
      const actualNonce = state.pendingTransactions
        .filter(tnx => tnx.state === 'pending')
        .map(tnx => tnx.nonce)
        .sort((a, b) => (a > b ? 1 : -1))
        .reduce((lastNonce, tnxNonce) => {
          if (lastNonce === tnxNonce) {
            return +tnxNonce + 1;
          }
          return lastNonce;
        }, nonce.toString())
        .toString();

      return actualNonce;
    },
    async signTransaction({ rootState }, { transaction, password }) {
      const { wallet } = rootState.accounts;
      const signedTransaction = await wallet.signTransaction(
        transaction,
        password,
      );

      return signedTransaction;
    },
    async sendSignedTransaction(
      { rootState, dispatch },
      { transaction, password },
    ) {
      const { wallet } = rootState.accounts;
      const { eth } = web3;

      try {
        // let hash;

        if (!transaction.nonce) {
          const nonce = await dispatch('getNextNonce');
          Object.assign(transaction, { nonce });
        }

        const tx = new Tx(transaction.getApiObject(eth));
        const signedTx = await dispatch('signTransaction', {
          transaction: tx,
          password,
        });
        const serializedTx = signedTx.serialize();
        const preparedTrx = `0x${serializedTx.toString('hex')}`;
        const eventEmitter = new EventEmitter();
        const sendEvent = eth
          .sendSignedTransaction(preparedTrx)
          .once('transactionHash', trxHash => {
            // hash = trxHash;
            eventEmitter.emit('transactionHash', trxHash);
          })
          .once('error', (err, receipt) => {
            const ignoreErrors = [
              'Transaction ran out of gas',
              'Transaction was not mined within750 seconds',
            ];
            const errInx = ignoreErrors.findIndex(errMsg =>
              err.message.includes(errMsg),
            );
            const isOutGas = errInx === 0 && receipt;

            if (errInx > 0 || isOutGas) {
              dispatch('handleSendingError', { err, receipt, transaction });
              eventEmitter.emit('error', err);
            }
          })
          .then(() => {
            eventEmitter.emit('confirmation');
          });
        // .catch(async ({ message }, receipt) => {
        //   if (message.includes('Transaction ran out of gas') && !receipt) {
        //     const interval = setInterval(async () => {
        //       const trx = await web3.eth.getTransactionReceipt(hash);

        //       if (trx && trx.status === true) {
        //         clearInterval(interval);
        //         eventEmitter.emit('confirmation');
        //       }
        //     }, 5000);
        //   }
        // });

        return eventEmitter;
      } catch (err) {
        dispatch('handleSendingError', { transaction });
      }
    },
    sendTransaction({ dispatch, commit }, { transaction, password }) {
      return dispatch('sendSignedTransaction', { transaction, password }).then(
        sendEvent =>
          new Promise((res, rej) => {
            sendEvent.once('transactionHash', hash => {
              // transaction.state = 'pending';
              // transaction.hash = hash;
              // transaction.date = new Date();
              Object.assign(transaction, {
                state: 'pending',
                hash,
                date: new Date(),
              });
              commit('addTransaction', transaction);
              res(hash);
            });

            sendEvent.once('confirmation', () => {
              console.log('sendEvent.once(confirmation)');
              const { hash } = transaction;
              const payload = { state: 'success' };

              commit('updateTransaction', {
                hash,
                payload,
              });
            });

            sendEvent.once('error', error => {
              const { hash } = transaction;
              const payload = { state: 'error', error };

              commit('updateTransaction', {
                hash,
                payload,
              });
              rej();
            });
          }),
      );
    },
    resendTransaction({ dispatch, state, commit }, { transaction, password }) {
      return dispatch('sendSignedTransaction', { transaction, password }).then(
        sendEvent =>
          new Promise((res, rej) => {
            const trxInList = state.pendingTransactions.find(
              trx => transaction.hash === trx.hash,
            );
            const { hash } = trxInList;

            sendEvent.once('transactionHash', newHash => {
              commit('updateTransaction', {
                hash,
                payload: { hash: newHash },
              });
              res(hash);
            });

            sendEvent.once('confirmation', () => {
              commit('updateTransaction', {
                hash,
                payload: { state: 'success' },
              });
            });

            sendEvent.once('error', error => {
              commit('updateTransaction', {
                hash,
                payload: { state: 'error', error },
              });
              rej();
            });
          }),
      );
    },
    cancelTransaction({ state, dispatch, commit }, { transaction, password }) {
      return dispatch('sendSignedTransaction', { transaction, password }).then(
        sendEvent =>
          new Promise((res, rej) => {
            const trxInList = state.pendingTransactions.find(
              trx => transaction.hash === trx.hash,
            );
            const { hash } = trxInList;

            sendEvent.once('transactionHash', async () => {
              const nonceInBlock = await dispatch('getNonceInBlock');

              if (trxInList.nonce == nonceInBlock) {
                const shortTnx = trxInList.hash.slice(0, 10);
                const error = new NotificationError({
                  title: 'Try to cancel the transaction',
                  text: `The cancellation ${shortTnx}... was started`,
                });
                dispatch('errors/emitError', error, { root: true });
              } else {
                // if a transaction with an too high nonce is canceled
                sendEvent.emit('confirmation');
              }
            });

            sendEvent.once('confirmation', () => {
              commit('updateTransaction', {
                hash,
                payload: { state: 'canceled' },
              });
              res();
            });

            sendEvent.once('error', error => {
              commit('updateTransaction', {
                hash,
                payload: { state: 'error', error },
              });
              rej();
            });
          }),
      );
    },
    handleSendingError(
      { dispatch },
      { err = {}, receipt, transaction = {} } = {},
    ) {
      const { err: errorMessage = '' } = err;
      const cause =
        receipt || errorMessage.includes('out of gas')
          ? ', because out of gas'
          : '';
      const { hash } = transaction;
      const shortHash = hash ? `${hash.slice(0, 4)}...${hash.slice(-4)} ` : '';
      const error = new NotificationError({
        title: 'Error sending transaction',
        text: `Transaction ${shortHash}was not sent${cause}`,
        type: 'is-danger',
      });
      dispatch('errors/emitError', error, { root: true });
    },
    // Transaction history from ethplorer
    async getTransactionHistory({ commit, dispatch, rootState }) {
      if (!rootState.accounts.address) return;

      try {
        const address = rootState.accounts.address.getChecksumAddressString();
        const [transactions, history] = await Promise.all([
          ethplorerService.getInfo(address),
          ethplorerService.getHistory(address),
        ]);
        const allTrx = transactions
          .concat(history)
          .map(trx => new Transaction(trx));

        commit('setTransactionHistory', allTrx);
        dispatch(
          'connectionStatus/updateApiErrorStatus',
          {
            id: 'ethplorer',
            status: true,
          },
          { root: true },
        );
      } catch (e) {
        const error = new NotificationError({
          ...e,
          title: 'Failed to get transaction information',
          text:
            'An error occurred while retrieving transaction information. Please try again.',
          type: 'is-warning',
          apiError: {
            id: 'ethplorer',
            status: false,
          },
        });
        dispatch('errors/emitError', error, { root: true });
      }
    },
    // Show notification of incoming transactions from block
    handleBlockTransactions({ dispatch, rootState }, transactions) {
      const { wallets } = rootState.accounts;
      const userAddresses = Object.keys(wallets).map(wallet =>
        wallet.toLowerCase(),
      );
      const toUserTrx = transactions.filter(trx =>
        userAddresses.some(address => address === trx.to),
      );

      toUserTrx.forEach(trx => {
        const { hash, to } = trx;
        const shortAddress = `${to.slice(0, 4)}...${to.slice(-4)}`;
        const shortHash = `${hash.slice(0, 4)}...${hash.slice(-4)}`;
        const error = new NotificationError({
          title: 'Incoming transaction',
          text: `Address ${shortAddress} received transaction ${shortHash}`,
        });

        dispatch('errors/emitError', error, { root: true });
      });

      if (!rootState.accounts.address) {
        return;
      }

      const address = rootState.accounts.address.getAddressString();
      const trxAddresses = toUserTrx.map(({ to }) => to);
      const { id: currentNetID } = rootState.web3.activeNet;

      if (
        currentNetID === MAIN_NET_ID &&
        trxAddresses.some(trxAddress => trxAddress === address)
      ) {
        dispatch('getTransactionHistory');
      }
    },
  },
};
