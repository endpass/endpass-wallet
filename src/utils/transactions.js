import { toBuffer } from 'ethereumjs-util';
import { decodeTx } from 'ethereum-tx-decoder';

/**
 * Converts transaction to rlp hash
 * @param {Tx} tx Transaction object
 * @returns {String} Transaction rlp hash
 */
export const convertTransactionToHash = tx =>
  `0x${tx.serialize().toString('hex')}`;

/**
 * Restores transaction from rlp hash
 * @param {String} rlpHash Rlp transaction hash
 * @returns {Object} Transaction object
 */
export const recoverTransaction = rlpHash => decodeTx(toBuffer(rlpHash));
