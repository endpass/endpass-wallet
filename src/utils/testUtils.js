import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
