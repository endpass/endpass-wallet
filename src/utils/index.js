import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
