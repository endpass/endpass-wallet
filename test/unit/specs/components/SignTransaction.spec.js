import Vuex from 'vuex';
import { ethplorerTransactions } from 'fixtures/transactions';
import { mount, createLocalVue } from '@vue/test-utils';
import Notifications from 'vue-notification';
import SignTransaction from '@/components/SignTransaction';
import { generateStubs } from '@/utils/testUtils';

describe('SignTransaction', () => {
  let wrapper;

  beforeEach(() => {
    const localVue = createLocalVue();
    const store = new Vuex.Store({
      state: {
        accounts: {
          wallet: {
            getPrivateKey: jest.fn(() => 'private key'),
          },
        },
        web3: {},
      },
    });

    localVue.use(Notifications);

    wrapper = mount(SignTransaction, {
      store,
      localVue,
      stubs: generateStubs(SignTransaction),
    });
  });

  describe('render', () => {
    it('should render the initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should enable "Sign transaction" button', () => {
      wrapper.setData({
        transaction: JSON.stringify(ethplorerTransactions[0]),
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('should show signed message', () => {
      wrapper.setData({
        signedTransaction: '0x0',
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('should show modal window for password confirmation', () => {
      wrapper.setData({
        transaction: JSON.stringify(ethplorerTransactions[0]),
        isPasswordModal: true,
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('signTransaction', () => {
      const password = 'password';

      beforeEach(() => {
        spyOn(wrapper.vm, '$notify');
      });

      it('should sign transaction', async () => {
        expect.assertions(3);

        wrapper.vm.wallet.signTransaction = jest
          .fn()
          .mockResolvedValueOnce('0x0');
        wrapper.setData({
          transaction: '{}',
        });

        await wrapper.vm.signTransaction(password);

        expect(wrapper.vm.signedTransaction).toEqual('0x0');
        expect(wrapper.vm.wallet.signTransaction).toBeCalled();
        expect(wrapper.vm.$notify).not.toHaveBeenCalled();
      });

      it('should not sign message', async () => {
        expect.assertions(3);

        wrapper.vm.wallet.signTransaction = jest.fn(() => {
          throw new Error();
        });
        global.console.error = jest.fn();

        await wrapper.vm.signTransaction(password);

        expect(wrapper.vm.signedTransaction).toBeNull();
        expect(wrapper.vm.$notify).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$notify).toHaveBeenCalledWith({
          title: 'Error signing transaction',
          text:
            'An error occurred while signing the transaction. Please try again.',
          type: 'is-danger',
        });
      });
    });
  });
});
