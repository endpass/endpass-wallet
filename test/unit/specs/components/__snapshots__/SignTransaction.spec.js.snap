// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignTransaction render should enable "Sign transaction" button 1`] = `
<div>
  <v-form
    id="signing-transaction-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      label="Transaction object (JSON)"
      placeholder="This is a transaction that you are signing to prove that you own it."
      v-model="transaction"
    />
     
    <v-button
      class-name="is-primary is-medium"
    >
      Sign transaction
    </v-button>
     
    <!---->
  </v-form>
   
  <!---->
</div>
`;

exports[`SignTransaction render should render the initial state of the component 1`] = `
<div>
  <v-form
    id="signing-transaction-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      label="Transaction object (JSON)"
      placeholder="This is a transaction that you are signing to prove that you own it."
      v-model="transaction"
    />
     
    <v-button
      class-name="is-primary is-medium"
      disabled="true"
    >
      Sign transaction
    </v-button>
     
    <!---->
  </v-form>
   
  <!---->
</div>
`;

exports[`SignTransaction render should show modal window for password confirmation 1`] = `
<div>
  <v-form
    id="signing-transaction-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      label="Transaction object (JSON)"
      placeholder="This is a transaction that you are signing to prove that you own it."
      v-model="transaction"
    />
     
    <v-button
      class-name="is-primary is-medium"
    >
      Sign transaction
    </v-button>
     
    <!---->
  </v-form>
   
  <code-password-modal
    code="{\\"timestamp\\":1535035155,\\"from\\":\\"0xb14ab53e38da1c172f877dbc6d65e4a1b0474c3c\\",\\"to\\":\\"0x9ECEEfDF3554E178a6549006F2C02163E63C9fd8\\",\\"hash\\":\\"0x53d26efcde07f1b2b68f3e1de93b730deabd1094970d9a68efb799e048e00892\\",\\"value\\":0.048671,\\"input\\":\\"0x\\",\\"success\\":true}"
    label="Transaction object"
    v-on:close="bound togglePasswordModal"
    v-on:confirm="bound signTransaction"
  />
</div>
`;

exports[`SignTransaction render should show signed message 1`] = `
<div>
  <v-form
    id="signing-transaction-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      label="Transaction object (JSON)"
      placeholder="This is a transaction that you are signing to prove that you own it."
      v-model="transaction"
    />
     
    <v-button
      class-name="is-primary is-medium"
      disabled="true"
    >
      Sign transaction
    </v-button>
     
    <div
      class="field"
    >
      <label
        class="label"
      >
        RLP encoded transaction
      </label>
       
      <p
        class="code"
      >
        0x0
      </p>
    </div>
  </v-form>
   
  <!---->
</div>
`;
